### ISSUE TITLE
Local redirect policy stops working after pod restart on EKS kiam

### ISSUE THREAD

**Issue by enkov** [cilium/cilium#16920]
## Bug report

**General Information**

- Cilium version (run `cilium version`)
```
Client: 1.10.2 319d2ac 2021-07-02T03:00:06+02:00 go version go1.16.5 linux/amd64
Daemon: 1.10.2 319d2ac 2021-07-02T03:00:06+02:00 go version go1.16.5 linux/amd64
```
- Kernel version (run `uname -a`)
```
Linux ip-172-29-65-4 5.4.0-1049-aws #51-Ubuntu SMP Wed May 12 21:13:52 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```
- Orchestration system version in use (e.g. `kubectl version`, ...)
```
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.3", GitCommit:"ca643a4d1f7bfe34773c74f79527be4afd95bf39", GitTreeState:"clean", BuildDate:"2021-07-15T20:58:09Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.2", GitCommit:"092fbfbf53427de67cac1e9fa54aaa09a28371d7", GitTreeState:"clean", BuildDate:"2021-06-16T12:53:14Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"linux/amd64"}
```
- Link to relevant artifacts (policies, deployments scripts, ...)
- Generate and upload a system zip:
```
curl -sLO https://git.io/cilium-sysdump-latest.zip && python cilium-sysdump-latest.zip
```
https://drive.google.com/file/d/17OF-b-QGP9k0MyC5lnmwg1VEVQSObr6D/view?usp=sharing

**How to reproduce the issue**

1. Deploy kiam and check that where is no lrp

```
root@ip-172-29-65-4:/home/cilium# cilium lrp list
LRP namespace   LRP name   FrontendType   Matching Service
```

2. Deploy lrp

```
root@ip-172-29-65-4:/home/cilium# cilium lrp list
LRP namespace   LRP name        FrontendType   Matching Service
kiam            kiam-redirect   IP + port
                |               169.254.169.254:80/TCP -> 172.29.65.4:8181(kiam/awsstaging-kiam-agent-s6d65),
```

3. Delete pod kiam/awsstaging-kiam-agent-s6d65 and check that new pod up and running

```
k get po -n kiam -o wide
awsstaging-kiam-agent-5nrf5    1/1     Running   0          19s     172.29.65.4     ip-172-29-65-4.ec2.internal
```

4. Check lrp from cilium agent on this node

```
root@ip-172-29-65-4:/home/cilium# cilium lrp list
LRP namespace   LRP name        FrontendType   Matching Service
kiam            kiam-redirect   IP + port
                |               169.254.169.254:80/TCP ->
```

As you can see now cilium doesn't find new pod

Lrp policy I use

```
apiVersion: "cilium.io/v2"
kind: CiliumLocalRedirectPolicy
metadata:
  name: "kiam-redirect"
  namespace: kiam
spec:
  redirectFrontend:
    addressMatcher:
      ip: "169.254.169.254"
      toPorts:
        - port: "80"
          protocol: TCP
  redirectBackend:
    localEndpointSelector:
      matchLabels:
        app: kiam
        component: "agent"
    toPorts:
      - port: "8181"
        protocol: TCP
```

Kiam installed from chart `uswitch/kiam --version 6.1.0` with this custom vars

```
agent:
  allowRouteRegexp: .*
  host:
    iptables: false
  log:
    level: info
  prometheus:
    servicemonitor:
      enabled: true
server:
  log:
    level: info
  nodeSelector:
    node-role.kubernetes.io/control-plane: ""
  prometheus:
    servicemonitor:
      enabled: true
  tolerations:
  - operator: Exists
```

**Comment by aditighag** [comment]
@enkov I found a regression in the code while trying to reproduce this issue. I've tested the [fix](https://github.com/cilium/cilium/pull/18563) on an EKS setup configured with kiam. Let me know if that resolves your issue.

**Comment by enkov** [comment]
Hello! Thank you for the fix.
Right now I don't have a cluster to test it.
Can you tell me in which version of cilium this fix will go?

**Comment by aditighag** [comment]
@enkov The fix will be backported and released in the next 1.11.x and 1.10.x versions.
